execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -256,34 +256,61 @@ where
             continue;
         }
-        let maybe_retrieved_trie: Option<Trie<K, V>> = store.get(txn, &trie_key)?;
+        let maybe_retrieved_trie_bytes: Option<Bytes> = store.get_raw(txn, &trie_key)?;
         // Perform an optional integrity check on the retrieved node.
         if check_integrity {
-            if let Some(trie_value) = &maybe_retrieved_trie {
-                let hash_of_trie_value = {
-                    let node_bytes = trie_value.to_bytes()?;
-                    Digest::hash(&node_bytes)
-                };
+            if let Some(trie_value_bytes) = &maybe_retrieved_trie_bytes {
+                let hash_of_trie_value = Digest::hash(trie_value_bytes);
                 if trie_key != hash_of_trie_value {
                     warn!(
-                        "Trie key {:?} has corrupted value {:?} (hash of value is {:?}); \
-                     adding to list of missing nodes",
-                        trie_key, trie_value, hash_of_trie_value
+                        "Trie key {:?} has corrupted value with hash {:?}; adding to list of missing nodes",
+                        trie_key, hash_of_trie_value,
                     );
                     missing_descendants.push(trie_key);
                     continue;
                 }
             }
         }
-        match maybe_retrieved_trie {
-            // If we can't find the trie_key; it is missing and we'll return it
+
+        // Optimization: Don't look for descendants of leaves.
+        match maybe_retrieved_trie_bytes
+            .as_ref()
+            .and_then(|bytes| bytes.get(0))
+        {
+            // if this is a leaf don't parse (ie, the first byte is 0), just continue
+            Some(0) => continue,
+            // Either no entry under trie key or data is empty.  If data is empty then the trie is
+            // corrupted. Either way treat as missing.
             None => {
                 missing_descendants.push(trie_key);
+                continue;
             }
-            // If we could retrieve the node and it is a leaf, the search can move on
-            Some(Trie::Leaf { .. }) => (),
+            _ => (),
+        }
+
+        // Parse the trie, handling errors gracefully.
+        let retrieved_trie: Trie<K, V> = match maybe_retrieved_trie_bytes
+            .map(bytesrepr::deserialize_from_slice)
+            .transpose()
+        {
+            Ok(Some(retrieved_trie)) => retrieved_trie,
+            Ok(None) => {
+                missing_descendants.push(trie_key);
+                continue;
+            }
+            // Couldn't parse; treat as missing and continue
+            Err(err) => {
+                error!(?err, "unable to parse trie");
+                missing_descendants.push(trie_key);
+                continue;
+            }
+        };
+
+        match retrieved_trie {
+            // Should be unreachable due to checking the first byte as a shortcut above.
+            Trie::Leaf { .. } => (),
```

Comment:
Consider adding a `warn!`, because if something should be unreachable but we somehow reach it, we should be able to detect that.

Reply:
Adding it here and above.

Reply:
5f072c245